packages/grafana-data/src/utils/vector.ts

```diff
@@ -109,18 +99,106 @@ export class CircularVector<T = any> implements Vector<T> {
   }
 }
 
+interface CircularOptions<T> {
+  buffer?: T[];
+  append?: 'head' | 'tail';
+  capacity?: number;
+}
+
 /**
- * Values are returned in the order defined by the input parameter
+ * Circular vector uses a single buffer to capture a stream of values
+ * overwriting the oldest value on add.
+ *
+ * This supports addting to the 'head' or 'tail' and will grow the buffer
+ * to match a configured capacity.
  */
-export class SortedVector<T = any> implements Vector<T> {
-  constructor(private source: Vector<T>, private order: number[]) {}
+export class CircularVector<T = any> implements Vector<T> {
+  private buffer: T[];
+  private index: number;
+  private capacity: number;
+  private tail: boolean;
+
+  constructor(options: CircularOptions<T>) {
+    this.buffer = options.buffer || [];
+    this.capacity = this.buffer.length;
+    this.tail = 'head' !== options.append;
+    this.index = 0;
 
-  get length(): number {
-    return this.source.length;
+    this.add = this.getAddFunction();
+    if (options.capacity) {
+      this.setCapacity(options.capacity);
+    }
   }
 
-  get(index: number): T {
-    return this.source.get(this.order[index]);
+  // Pick an add function that either writes forwards or backwards
+  private getAddFunction() {
+    if (this.tail) {
+      return (value: T) => {
+        this.buffer[this.index] = value;
+        this.index = (this.index + 1) % this.buffer.length;
+      };
+    }
+
+    // Append values to the head
+    return (value: T) => {
+      let idx = this.index - 1;
```

Comment:
Do we actually need different algo for this. Could we fill it the same way and just return the data reversed afterwards?

Reply:
What is the access api? Is it only toArray and get? That should be simple to reverse it there or am I missing something?

Reply:
> this version adds them in a different order and gets them in the same order. 6 vs 1/2 dozen?

Yeah maybe, seemed to me like you could do something like:
```
toArray:
   return this.tail ? reversed(this.buffer) : this.buffer

get(i):
  return this.tail ? reversed(this.buffer[i]) : this.buffer[i]
```

but maybe I am missing something.